### PR TITLE
HIVE-25582: Empty result when using offset limit with MR

### PR DESCRIPTION
**Pull request by dengzhhu653**

### CODE REVIEW

**Comment by kgyrtkirk** [comment]
Wouldn't it be possible to somehow fix the default cache implementation instead of hacking that condition with !tez further? 
And from the above comments I suspect that this could be covered with a test - or it needs some special circumstances?

**Comment by dengzhhu653** [comment]
> Wouldn't it be possible to somehow fix the default cache implementation instead of hacking that condition with !tez further? And from the above comments I suspect that this could be covered with a test - or it needs some special circumstances?

Thanks a lot for the feedback! Let me take a look at the default cache implementation, and add a qtest to cover this.

**Comment by dengzhhu653** [comment]
The failed test seems unrelated...

File: ql/src/java/org/apache/hadoop/hive/ql/exec/mr/ObjectCache.java [hunk 1]
```diff
@@ -18,82 +18,66 @@
...
-        return value;
-      }
-
-      @Override
-      public T get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException,
-          TimeoutException {
-        return value;
-      }
-    };
+    return staticPool.submit((Callable)() -> retrieve(key, fn));
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/mr/ObjectCache.java:76]: I think the old `new Future` was somewhat better - as it not needed an "executor" to be running in the background to service this method

Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/mr/ObjectCache.java]: Done

**Review by kgyrtkirk — APPROVED** [review]

**Comment by dengzhhu653** [comment]
Thanks a lot for the feedback, review and merge :)

**Pull request merged**